common-security/src/main/java/co/cask/common/security/authorization/client/DefaultAuthorizationClient.java

```diff
@@ -0,0 +1,119 @@
+ * the License.
+ */
+package co.cask.common.security.authorization.client;
+
+import co.cask.common.security.authorization.ACLEntry;
+import co.cask.common.security.authorization.ACLStore;
+import co.cask.common.security.authorization.AuthorizationContext;
+import co.cask.common.security.authorization.IdentifiableObject;
+import co.cask.common.security.authorization.ObjectId;
+import co.cask.common.security.authorization.SubjectId;
+import co.cask.common.security.authorization.UnauthorizedException;
+import com.google.common.base.Predicate;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Sets;
+import com.google.inject.Inject;
+
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Default implementation of {@link AuthorizationClient} which uses {@link ACLStore}.
+ */
+public class DefaultAuthorizationClient implements AuthorizationClient {
+
+  private final ACLStore aclStore;
+  private final AuthorizationContext context;
+
+  /**
+   * @param aclStore used to set and get {@link ACLEntry}s
+   * @param context provides current user and its groups
+   */
+  @Inject
+  public DefaultAuthorizationClient(ACLStore aclStore, AuthorizationContext context) {
+    this.aclStore = aclStore;
+    this.context = context;
+  }
+
+  @Override
+  public void authorize(Iterable<? extends ObjectId> objects, Iterable<? extends SubjectId> subjects,
+                        Iterable<String> requiredPermissions) throws UnauthorizedException {
+    if (!isAuthorized(objects, subjects, requiredPermissions)) {
+      throw new UnauthorizedException();
+    }
+  }
+
+  @Override
+  public void authorize(ObjectId object, Iterable<? extends SubjectId> subjects,
+                        Iterable<String> requiredPermissions) throws UnauthorizedException {
+    if (!isAuthorized(object, subjects, requiredPermissions)) {
+      throw new UnauthorizedException();
+    }
+  }
+
+  @Override
+  public void authorizeCurrentUser(ObjectId object,
+                                   Iterable<String> requiredPermissions) throws UnauthorizedException {
+
+    Iterable<? extends SubjectId> subjects = Iterables.concat(context.getCurrentUsersGroups(),
+                                                              ImmutableSet.of(context.getCurrentUser()));
+    if (!isAuthorized(object, subjects, requiredPermissions)) {
+      throw new UnauthorizedException();
+    }
+  }
+
+  @Override
+  public boolean isAuthorized(Iterable<? extends ObjectId> objects, Iterable<? extends SubjectId> subjects,
+                              Iterable<String> requiredPermissions) {
+
+    Set<String> remainingRequiredPermission = Sets.newHashSet(requiredPermissions);
+    for (ObjectId object : objects) {
+      for (SubjectId subject : subjects) {
+        for (ACLEntry aclEntry : aclStore.read(new ACLStore.Query(object, subject))) {
```

Comment:
In most cases, probably yes, since `aclStore.read` would usually be talking directly to HBase or a SQL server. Probably should add a way to query for multiple objects and subjects.